### PR TITLE
feat: Add waiter for object

### PR DESCRIPTION
**Pull request by dlipovetsky**
**What problem does this PR solve?**:
Implements a wait for a check to pass against a typed object.

We'll use this in some lifecycle handers, e.g., in a future change for deploying ServiceLoadBalancer configuration to the remote cluster.

**Which issue(s) this PR fixes**:
Fixes #

**How Has This Been Tested?**:

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/wait/wait_test.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+						ObjectMeta: v1.ObjectMeta{
+							Name: "example",
+						},
+					},
+				},
+			},
+			errCheck: func(err error) bool {
+				return errors.Is(err, &CheckFailedError{}) &&
+					!wait.Interrupted(err)
+			},
```
Comment (dkoshkin) [inline pkg/wait/wait_test.go:123]: Nice!

**Review by dkoshkin — APPROVED** [review]

File: pkg/wait/wait.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+	var getErr error
+	waitErr := wait.PollUntilContextTimeout(
+		ctx,
+		input.Interval,
+		input.Timeout,
+		true,
+		func(checkCtx context.Context) (bool, error) {
+			if getErr = input.Reader.Get(checkCtx, key, input.Target); getErr != nil {
+				// Retry if get fails.
```
Comment (jimmidyson) [inline pkg/wait/wait.go:54]: Maybe only try on not found?

File: pkg/wait/wait.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+				return false, &CheckFailedError{cause: err}
+			} else {
+				// Retry if check fails.
+				return ok, nil
+			}
+		})
+
+	if wait.Interrupted(waitErr) {
+		if getErr != nil {
+			return fmt.Errorf("%w; last get error: %w", waitErr, getErr)
```
Comment (jimmidyson) [inline pkg/wait/wait.go:68]: Can you use multiple wrapped errors like this? Do both get included in error chain?

Comment (jimmidyson) [reply pkg/wait/wait.go:68]: From https://pkg.go.dev/fmt#Errorf

> If there is more than one %w verb, the returned error will implement an Unwrap method returning a []error containing all the %w operands in the order they appear in the arguments.

So all good here! TIL :)

**Review by jimmidyson — APPROVED** [review]

**Review by faiq — APPROVED** [review]

**Pull request closed without merging**